server/app/tests/test_user.py

```diff
@@ -58,4 +66,151 @@ def test_change_password_new_password_is_same_as_current_password(user_service,
     mock_user_repo.find_by_id.return_value = user_start
 
     with pytest.raises(UserException) as e:
-        user_service.change_password(1, dto)
\ No newline at end of file
+        user_service.change_password(1, dto)
+
+def test_add(user_service, mock_user_repo):
+    dto = UserRegisterDTO(username="Username1", email="email1@email.com", password="Password1")
+    mock_user_repo.find_by_email.return_value = None
+    mock_user_repo.find_by_username.return_value = None
+
+    test_result = user_service.add(dto)
+
+    assert test_result.id is not None
+
+def test_add_existing_user(user_service, mock_user_repo):
+    dto = UserRegisterDTO(username="Username1", email="email1@email.com", password="Password1")
+    user = User(username="Username1", email="email1@email.com", role=UserRole.user, hashed_password=hash_password(dto.password))
+    
+    def get_user_by_email_mocked(email: str):
+        return user if user.email == email else None
+    mock_user_repo.find_by_email.side_effect = get_user_by_email_mocked
+
+    def get_user_by_username_mocked(username: str):
+        return user if user.username == username else None
+    mock_user_repo.find_by_username.side_effect = get_user_by_username_mocked
+
+    # [1] Creating a user with a non-unique email
+    with pytest.raises(FieldTakenException) as e:
+        user_service.add(dto)
+    assert e.value.message == "Email already taken"
+
+    # [2] Creating a user with a non-unique username
+    dto.email = "email2@email.com"
+    with pytest.raises(FieldTakenException) as e:
+        user_service.add(dto)
+    assert e.value.message == "Username already taken"
+
+def test_add_admin(user_service, mock_user_repo):
+    dto = UserRegisterDTO(username="Username1", email="a@email.com", password="Password1")
+    user_start = User(id=1, username="Username1", email="a@email.com", role=UserRole.user, hashed_password=hash_password("Password1"))
+    user_end = User(id=1, username="Username1", email="a@email.com", role=UserRole.admin, hashed_password=hash_password("Password1"))
+
+    with mock.patch.object(user_service, "add", return_value=user_start) as _:
+        mock_user_repo.set_role.return_value = user_end
+
+        def set_role(user: User, role: UserRole):
+            user.sqlmodel_update({"role": role})
+            return user
+        mock_user_repo.set_role.side_effect = set_role
+
+        result = user_service.add_admin(dto)
+        
+        assert result.role == UserRole.admin
```

Comment:
You don't have to (shouldn't?) mock `add()` here.

When unit testing, you mock _external dependencies_. `user_repo` is an external dependency which is why we mock it. When testing a class, you don't mock other parts of the class because those are _implementation details_.

Should you decide to refactor your code tomorrow (e.g. have `add_admin` not rely on `add` anymore), the test won't pass anymore. But the whole point of refactoring is to change the structure of the code without changing its behavior, so the test _must_ pass.

The other reason is that we _want_ `add_admin` to fail when `add` fails. We've discussed whether you should test `add_admin` as thoroughly as `add`, but there's really no need for such trivial code. This unit test cares about input (parameters) and output (return value).

There's a good [thread](https://stackoverflow.com/questions/57235607/how-to-mock-a-method-that-calls-other-methods-that-are-present-in-the-same-class) on StackOverflow discussing this.